File: cpp/104/solution2.cpp
#include <queue>


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};


class Solution {
    public:
        int maxDepth(TreeNode *root) {
            std::queue<TreeNode*> queue;
            if (root) queue.push(root);
            int ans = 0;
            while (!queue.empty()) {
                int sz = queue.size();
                for (int i = 0; i < sz; i++) {
                    TreeNode *cur = queue.front();
                    queue.pop();
                    if (cur->left) queue.push(cur->left);
                    if (cur->right) queue.push(cur->right);
                }
                ++ans;
            }
            return ans;
        }
};

File: java/104/Solution2.java
import java.util.LinkedList;
import java.util.Queue;

class Solution2 {
	public int maxDepth(TreeNode root) {
		if (root == null) {
			return 0;
		}

		Queue<TreeNode> q = new LinkedList<>();
		q.offer(root);

		int depth = 0;
		while (!q.isEmpty()) {
			int size = q.size();
			while (size-- > 0) {
				TreeNode node = q.poll();
				if (node.left != null) {
					q.offer(node.left);
				}
				if (node.right != null) {
					q.offer(node.right);
				}
			}
			depth++;
		}
		return depth;
	}
}
